src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -63,25 +63,31 @@ public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilatio
 
             // run the actual generation
             var state = StateWithPendingEditsApplied(_state);
-            var sourcesBuilder = PooledDictionary<ISourceGenerator, ImmutableArray<GeneratedSourceText>>.GetInstance();
+            var stateBuilder = PooledDictionary<ISourceGenerator, GeneratorState>.GetInstance();
 
             //PROTOTYPE: should be possible to parallelize this
             foreach (var generator in state.Generators)
             {
                 try
                 {
+                    // initialize the generator if needed
+                    GeneratorState generatorState;
+                    if (!state.GeneratorStates.TryGetValue(generator, out generatorState))
+                    {
+                        generatorState = InitializeGenerator(generator, cancellationToken);
+                    }
```

Comment:
Can this be moved ahead of the `try` block? It looks like `InitializeGenerator()` already catches exceptions.